File: snsmediadl/web/js/enums.js
// 前端用得到的值域，與 `db/enums.py` 對應。
//
// ⚠️ 加值時**三個地方都要改**：這裡、`db/enums.py`、以及一個 alembic migration
// （CHECK constraint 不會自己跟著 enum 改）。少改一處的症狀是「選了新選項存不進去」。
//
// 這一份被媒體詳情面板與帳號卡共用 —— 拆模組時它一度只留在 media.js，
// 結果帳號頁整片空白（`ReferenceError: opts is not defined`）。

export const RATINGS = ['', 'sfw', 'r18'];
// 要與 db/enums.py 的 ContentType 一致。加值時記得也要有 alembic migration ——
// CHECK constraint 不會自己跟著 enum 改。
export const CONTENTS = ['', 'illust', 'irl', 'mod', 'ai', '3d', 'photograph', 'other'];
// ⚠ 這裡原本還有一個 `opts()`，用來產生 `<option>` 字串。原生 `<select>`
// 全面退場（改用 dom.js 的 singleDrop / multiDrop）之後它沒有任何呼叫者，
// 已移除 —— 留著會讓下一個人以為「這裡還有 select 可以用」。
//
// ⚠ 上面的 RATINGS / CONTENTS **不要因為看起來沒人用就刪**：
// tests/test_enums_sync.py 直接用正規表示式抽這兩個陣列去跟 db/enums.py 對，
// 刪掉會讓那條「三份值域必須一致」的測試靜默失效。

// 多選篩選用的值域（不含空字串 —— 「不選」就是不選，沒有「全部」這個值）。
//
// ⚠️ `ugoira` 照列，即使 2026-08-15 的分布表沒有它。enum 有這個值就要能篩，
// 「看起來沒資料所以藏起來」會讓真的有資料的那天沒有人發現。
export const KINDS = ['photo', 'video', 'animated_gif', 'ugoira'];
export const RATING_VALUES = ['sfw', 'r18'];
export const CONTENT_VALUES = ['illust', 'irl', 'mod', 'ai', '3d', 'photograph', 'other'];

File: snsmediadl/db/enums.py
"""列舉值。

存成 TEXT + CHECK constraint，不用 SQLAlchemy native enum ——
SQLite 沒有原生 enum，用 native enum 之後改值要重建整張表。

Python 3.10 沒有 StrEnum，因此用 (str, Enum) 取得同樣的字串行為。
"""

from __future__ import annotations

from enum import Enum


class _Str(str, Enum):
    def __str__(self) -> str:  # pragma: no cover - 只是讓 f-string 好看
        return self.value

    @classmethod
    def values(cls) -> list[str]:
        return [m.value for m in cls]


class Platform(_Str):
    X = "x"
    MISSKEY = "misskey"
    BARAAG = "baraag"
    PIXIV = "pixiv"


class MediaKind(_Str):
    """媒體型別。掛在 media 不掛 post ——
    實測證實單一貼文可同時含 photo + video + animated_gif。"""

    PHOTO = "photo"
    VIDEO = "video"
    ANIMATED_GIF = "animated_gif"
    # pixiv 的動圖：一個 zip 裝一堆 jpg + 每幀的延遲毫秒數。
    # 第一版只存 zip 不轉檔（轉檔要拉 ffmpeg，是後處理不是採集）。
    UGOIRA = "ugoira"


class MediaStatus(_Str):
    PENDING = "pending"
    DOWNLOADING = "downloading"
    DONE = "done"
    FAILED = "failed"


class Rating(_Str):
    """分級。與 ContentType 正交 —— 「AI 生的 R18」是 ai + r18。

    NULL 代表「未知」，不是 sfw。猜錯的方向不對稱，未知就留空。
    """

    SFW = "sfw"
    R18 = "r18"


class ContentType(_Str):
    ILLUST = "illust"
    IRL = "irl"
    MOD = "mod"
    AI = "ai"
    # 成員名不能以數字開頭，所以是 THREE_D —— **值**才是 "3d"。
    # 存進 DB 與 API 的都是值，成員名只有 Python 這邊看得到。
    THREE_D = "3d"
    PHOTOGRAPH = "photograph"
    OTHER = "other"


class RatingSource(_Str):
    """分級是誰標的。不記來源的話，機器猜的會和人工確認的混在一起。"""

    MANUAL = "manual"
    ACCOUNT_DEFAULT = "account_default"
    AUTO = "auto"


class AccountRole(_Str):
    MAIN = "main"
    ALT = "alt"
    R18_ALT = "r18_alt"


class FetchStatus(_Str):
    """最後一次擷取的結果。

    `OK` 與 `NO_NEW` 刻意分開：兩者都是「成功」，但使用者要知道的是
    「有沒有東西進來」。合成一個 ok 就等於把答案藏起來 —— 而「查了但
    對方沒發新的」正是這個功能要消除的疑惑。
    """

    OK = "ok"                      # 跑完，有新貼文
    NO_NEW = "no_new"              # 跑完，沒有新的（增量立刻停）
    RATE_LIMITED = "rate_limited"  # 429
    NOT_FOUND = "not_found"        # 404 —— 帳號改名或打錯字
    AUTH_REQUIRED = "auth_required"  # 憑證缺失或過期
    FAILED = "failed"              # 其他例外
    SKIPPED = "skipped"            # 該站台已被標記限速，這輪先跳過
